File: src/lib/email/routing.ts
import { randomUUID } from 'node:crypto';

const ROUTING_TAG_PREFIX = 'c_';
const ROUTING_TOKEN_LENGTH = 32;
const TOKEN_PATTERN = /^[0-9a-f]{32}$/;

interface MailboxParts {
  local: string;
  domain: string;
}

function parseBaseMailbox(address: string): MailboxParts | null {
  const normalized = address.trim().toLowerCase();
  const at = normalized.lastIndexOf('@');
  if (
    at <= 0 ||
    at === normalized.length - 1 ||
    normalized.includes('<') ||
    normalized.includes('>') ||
    normalized.slice(0, at).includes('+')
  ) {
    return null;
  }

  const local = normalized.slice(0, at);
  const domain = normalized.slice(at + 1);
  const generatedLocalLength =
    local.length + 1 + ROUTING_TAG_PREFIX.length + ROUTING_TOKEN_LENGTH;
  const domainLabels = domain.split('.');
  if (
    !/^[a-z0-9.!#$%&'*/=?^_`{|}~-]+$/.test(local) ||
    local.startsWith('.') ||
    local.endsWith('.') ||
    local.includes('..') ||
    generatedLocalLength > 64 ||
    domain.length > 253 ||
    domainLabels.length < 2 ||
    domainLabels.some(
      (label) =>
        label.length === 0 ||
        label.length > 63 ||
        !/^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/.test(label),
    ) ||
    generatedLocalLength + domain.length + 1 > 254
  ) {
    return null;
  }
  return { local, domain };
}

export function isValidReplyToBaseAddress(address: string): boolean {
  return parseBaseMailbox(address) !== null;
}

export function createRoutingToken(): string {
  return randomUUID();
}

export function buildConversationReplyTo(
  baseAddress: string,
  routingToken: string,
): string {
  const base = parseBaseMailbox(baseAddress);
  const token = routingToken.replaceAll('-', '').toLowerCase();
  if (!base || !TOKEN_PATTERN.test(token)) {
    throw new Error('Invalid conversation Reply-To configuration');
  }
  return `${base.local}+${ROUTING_TAG_PREFIX}${token}@${base.domain}`;
}

export function extractRoutingTokens(
  addresses: readonly string[],
  baseAddress: string,
): string[] {
  const base = parseBaseMailbox(baseAddress);
  if (!base) {
    throw new Error('Invalid conversation Reply-To configuration');
  }

  const prefix = `${base.local}+${ROUTING_TAG_PREFIX}`;
  const tokens = new Set<string>();
  for (const rawAddress of addresses) {
    const address = rawAddress.trim().toLowerCase();
    const at = address.lastIndexOf('@');
    if (at <= 0 || address.slice(at + 1) !== base.domain) {
      continue;
    }
    const local = address.slice(0, at);
    if (!local.startsWith(prefix)) {
      continue;
    }
    const token = local.slice(prefix.length);
    if (TOKEN_PATTERN.test(token)) {
      tokens.add(
        `${token.slice(0, 8)}-${token.slice(8, 12)}-${token.slice(12, 16)}-${token.slice(16, 20)}-${token.slice(20)}`,
      );
    }
  }
  return [...tokens];
}

File: src/main/kotlin/com/castab/resend/email/Routing.kt
package com.castab.resend.email

import java.util.UUID

private const val ROUTING_TAG_PREFIX = "c_"
private const val ROUTING_TOKEN_LENGTH = 32
private val TOKEN_PATTERN = Regex("^[0-9a-f]{32}$")
private val LOCAL_PATTERN = Regex("^[a-z0-9.!#\$%&'*/=?^_`{|}~-]+$")
private val LABEL_PATTERN = Regex("^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$")

private data class MailboxParts(val local: String, val domain: String)

private fun parseBaseMailbox(address: String): MailboxParts? {
    val normalized = address.trim().lowercase()
    val at = normalized.lastIndexOf('@')
    if (at <= 0 ||
        at == normalized.length - 1 ||
        normalized.contains('<') ||
        normalized.contains('>') ||
        normalized.substring(0, at).contains('+')
    ) {
        return null
    }

    val local = normalized.substring(0, at)
    val domain = normalized.substring(at + 1)
    val generatedLocalLength = local.length + 1 + ROUTING_TAG_PREFIX.length + ROUTING_TOKEN_LENGTH
    val domainLabels = domain.split(".")
    if (!LOCAL_PATTERN.matches(local) ||
        local.startsWith(".") ||
        local.endsWith(".") ||
        local.contains("..") ||
        generatedLocalLength > 64 ||
        domain.length > 253 ||
        domainLabels.size < 2 ||
        domainLabels.any { it.isEmpty() || it.length > 63 || !LABEL_PATTERN.matches(it) } ||
        generatedLocalLength + domain.length + 1 > 254
    ) {
        return null
    }
    return MailboxParts(local, domain)
}

fun isValidReplyToBaseAddress(address: String): Boolean = parseBaseMailbox(address) != null

fun createRoutingToken(): String = UUID.randomUUID().toString()

fun buildConversationReplyTo(baseAddress: String, routingToken: String): String {
    val base = parseBaseMailbox(baseAddress)
    val token = routingToken.replace("-", "").lowercase()
    if (base == null || !TOKEN_PATTERN.matches(token)) {
        throw IllegalStateException("Invalid conversation Reply-To configuration")
    }
    return "${base.local}+$ROUTING_TAG_PREFIX$token@${base.domain}"
}

fun extractRoutingTokens(addresses: List<String>, baseAddress: String): List<String> {
    val base = parseBaseMailbox(baseAddress)
        ?: throw IllegalStateException("Invalid conversation Reply-To configuration")
    val prefix = "${base.local}+$ROUTING_TAG_PREFIX"
    val tokens = LinkedHashSet<String>()
    for (rawAddress in addresses) {
        val address = rawAddress.trim().lowercase()
        val at = address.lastIndexOf('@')
        if (at <= 0 || address.substring(at + 1) != base.domain) continue
        val local = address.substring(0, at)
        if (!local.startsWith(prefix)) continue
        val token = local.substring(prefix.length)
        if (TOKEN_PATTERN.matches(token)) {
            tokens.add(
                "${token.substring(0, 8)}-${token.substring(8, 12)}-${token.substring(12, 16)}-" +
                    "${token.substring(16, 20)}-${token.substring(20)}",
            )
        }
    }
    return tokens.toList()
}
